### PR TITLE
fix issue 23972 - class identity check is broken

### PR DESCRIPTION
**Pull request by ghost**
insert casts to Object when interfaces are used as IdentityExp operands

### CODE REVIEW

File: compiler/src/dmd/expressionsem.d [hunk 1]
```diff
@@ -12543,6 +12547,41 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+        {
+            auto tc1 = exp.e1.type.isTypeClass();
+            auto tc2 = exp.e2.type.isTypeClass();
+            if (tc1 && tc2)
+            {
+                const id1 = tc1 && !tc1.sym.isCPPinterface() ? tc1.sym.isInterfaceDeclaration() : null;
+                const id2 = tc2 && !tc2.sym.isCPPinterface() ? tc2.sym.isInterfaceDeclaration() : null;
+                if (id1 || id2)
+                {
+                    if (id1 && id2)
```
Comment (RazvanN7) [inline compiler/src/dmd/expressionsem.d:12564]: This looks weird. I would do it:

```d
if (id1)
{
    if (id2) { /* id1 && id2 */ }
    else { /* id1 */]
}
else if (id2)
{
    // id2
}
```
This gets rid of the extra condition of `if(id1 || id2)

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
I am not sure that the bug report is valid.

In this example:

```d
interface I { void g(); }
interface I1 : I { void g1(); }
interface I2 : I { void g2(); }
interface J : I1, I2 { void h(); }

class C : J
{
    override void g() { }
    override void g1() { }
    override void g2() { }
    override void h() { }
}

void main() @safe
{
    C c = new C;
    I i1 = cast(I1) c;
    I i2 = cast(I2) c;
```

I1 and I2 are two different interfaces, irrespective of the fact that an both i1 and i2 refer to the same object, I1 and I2 are different types.

I see this scenario similar to having an char* that is initialized to a part of an int*. Are char* and int* the same type if they are initialized to the same memory region? I don't think so.

**Comment by SixthDot** [comment]
> In this example:
> [...]

Indeed, the automatic cast introduces by this PR would give wrong result.
I'm gonna close this PR but I encourage you to explain that to the reporter
before clossing the issue as "invalid."

**Comment by dkorpel** [comment]
> irrespective of the fact that an both i1 and i2 refer to the same object, I1 and I2 are different types

This reasoning doesn't make sense to me, because the `is` operator doesn't compare types, it compares values for equal bits.

> Are char* and int* the same type if they are initialized to the same memory region? I don't think so.

You can't compare `char*` and `int*` with the `is` operator, but consider this example:
```D
void main() @safe
{
    int* x = null;
    void* y = x;
    assert(x is y); // passes
}
```

**Comment by SixthDot** [comment]
I'm not especialy pro or con the POV of the reporter so I reopen.

**Comment by SixthDot** [comment]
problem pointed in https://github.com/dlang/dmd/pull/15696#pullrequestreview-1682054176 is still valid however

**Comment by dkorpel** [comment]
The problem pointed out is "I1 and I2 are different types", and I pointed out "the is operator doesn't compare types", so I don't understand how it's "still valid".

**Comment by SixthDot** [comment]
So finally, the only case that is ok to handle is "iface is obj" ?

**Comment by PetarKirov** [comment]
The bug report is valid and the spec is also clear about that. This form of the [`is` expression](https://dlang.org/spec/expression.html#identity_expressions) compares the identities of two expressions and for reference types this means whether the two expressions refer to the same object.
This aspect of comparing references is similar to the [`Object.ReferenceEquals`](https://learn.microsoft.com/en-us/dotnet/api/system.object.referenceequals?view=net-7.0) method in C#, [`==` operator](https://docs.oracle.com/javase/specs/jls/se13/html/jls-15.html#jls-15.21.3) in Java or `===` and [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) in JavaScript.

> `is` should be a simple bit compare, and not be doing casting logic for classes.

@dkorpel I think you're deriving the semantic meaning of `a is b` expressions from an implementation detail (bit comparison), while the implementation strategy should follow (be derived) from the semantic meaning. So instead of saying "if you need to do bit comparisons use the `is` expression", we should say if you need to compare the identities if two expressions, use the `is` operator.

Or let's look at things from another perspective - let's say the `a is b` expression didn't exist in the language. If we were to add it today, which behavior would be more useful - comparing the pointer bit patterns of two reference variables, or actually checking if they point to the same object? Given that:
* one a can already check if two references have the same pointer bit pattern (by casting the references to pointers and then comparing them with the `==` operator)
* most mainstream OOP languages, share the same idea that references equality, identity comparisons and checking whether to variables point to the same object are all synonyms

I'd say that it's clear that:
* `i1 is c`
* `i2 is c`
* `i1 is i2`
(from the example in the bug report) should all be true.

**Comment by PetarKirov** [comment]
> I1 and I2 are two different interfaces, irrespective of the fact that an both i1 and i2 refer to the same object, I1 and I2 are different types.


@RazvanN7 I think you also confused the difference between static and dynamic type of expressions / variables. See, e.g.:
https://www.cs.cornell.edu/courses/JavaAndDS/files/staticType.pdf

> I see this scenario similar to having an char* that is initialized to a part of an int*. Are char* and int* the same type if they are initialized to the same memory region? I don't think so.

In this scenario, `int*` and `char*` clearly have different static types, even if they have the same value, i.e. point to the same memory address. I'm not sure if we can talk about dynamic types, as the concept of dynamic type relies on memory and type safety invariants enforced by the language, which is not the case for this scenario, while interfaces are object up and down casting are fully memory and type safe.

**Comment by dkorpel** [comment]
I don't know which interpretation is more useful in practice, but the idea of mathematical identity is that you can substitute identical elements in an expression and get the same result. So if `cast(void*) a - cast(void*) a == 0`, and `a is b == true`, then `cast(void*) a - cast(void*) b == 0` also, which would fail with `interface is class == true`

**Comment by dkorpel** [comment]
During the Nov 10 DLF monthly meeting, changing the specification was preferred over changing the behavior. A function like [Object.ReferenceEquals](https://learn.microsoft.com/en-us/dotnet/api/system.object.referenceequals?view=net-7.0) could still be added to druntime if desirable.

**Comment by SixthDot** [comment]
> changing the specification was preferred

I can't disagree. While implementing the fix was interesting there was actually a little something "not quite right" with the issue. Unlike many others, that was not a clear bug.

**Comment by PetarKirov** [comment]
> the idea of mathematical identity is that you can substitute identical elements in an expression and get the same result
Agreed.

> So if cast(void*) a - cast(void*) a == 0, and a is b == true, then cast(void*) a - cast(void*) b == 0 also, which would fail with interface is class == true

But why cast to `void*`? This conflates semantics with implementation details. We care if two expressions resolve to references to the same object, not whether the two pointers are equal. If the `extern(D)` class ABI changes the result may as well change. But ABI changes are not supposed to break high-level `@safe` D code.

Here's an implementation that preserves the mathematical equivalence properties:

Let:
```d


Given:
* `toObjAddr(a) - toObjAddr(a) == 0`
* `a is b == true`

are true, then:

* `toObjAddr(a) - toObjAddr(b) == 0` is also true

**Comment by PetarKirov** [comment]
(I somehow reopened the PR unintentionally while posting my comment.)

**Comment by dkorpel** [comment]
> But why cast to `void*`? 

Because then you get a counter example that shows the proposed `is` operator meaning is not equivalent to mathematical identity, unless you consider that cast 'undefined behavior' or something.

**Comment by dkorpel** [comment]
Note that casting to `void*` is not the only way to expose the pointer, consider an object/inferface inside a union. How is the new `is` operator supposed to deal with that?

**Comment by PetarKirov** [comment]
> Because then you get a counter example that shows the proposed `is` operator meaning is not equivalent to mathematical identity, unless you consider that cast 'undefined behavior' or something.

I don't get what you're saying. Can you explain in more detail? What counter example?

> Note that casting to `void*` is not the only way to expose the pointer [..]

Sure, but that's not what we're discussing here. I'm not sure why you're even bringing this up. If I work with pointers, I expect them to work like they do in C. If I work with class references, I expect them to work like they do in most OOP languages (of which probably C# is the closest to D, based on the `struct` vs class `dichotomy`). So if through whatever means I obtain a pointer to a class object (including interior pointer to a field or ptr to interface vtbl), I wouldn't expect the `is` operator to do anything but compare the address stored in the pointer. On the other hand, if the (static) types of the two sub-expressions in `<sub-expr-a> is <sub-expr-b>` are classes, the implementation should check if they point to the same object.

> consider an object/inferface inside a union. How is the new is operator supposed to deal with that?

It won't, of course. The operand types won't be of class type.

**Comment by dkorpel** [comment]
> I don't get what you're saying. Can you explain in more detail? What counter example?

You said I'm deriving the meaning of the `is` operator from an implementation detail, not mathematical identity.
My response is that the current `is` operator is closer to mathematical identity, while this PR's new `is` operator is not mathematical identity. The reasoning goes:

If `a` and `b` are identical, you can substitute `a` for `b` in *all* expressions without changing the result of the expression.
A universal quantification can be disproven with a single counter example, so I gave one example where substituting `a` for `b` would change the result, while `a is b == true` after this PR.
Therefore, `a is b` after this PR is not mathematical identity.

Note that it doesn't work the other way, showing one positive example (using your `toObjAddr` function) doesn't prove the universal quantification. The substitution rule has to hold for any expression.

Now you could argue that the `is` operator isn't mathematical identity to begin with because of compile time introspection (think `a.stringof != b.stringof` while `a is b`), or that strict mathematical identity isn't the goal of the `is` operator. Programming language design is subjective in the end. However, I'm not starting at 'the `is` operator has to do bit comparison' and deriving the semantic meaning from that, but the other way around.
 
> Sure, but that's not what we're discussing here. I'm not sure why you're even bringing this up.

Because you act like I'm cheating by casting to `void*`, while I consider it a valid expression just like any other. I'm bringing it up to show that even if we discount casting to `void*`, there's still other ways to break mathematical identity with the new `is` operator.

**Comment by PetarKirov** [comment]
Okay, I think I see where you're coming from. I think because of my poor choice of words in my first post:
> "This form of the [is expression](https://dlang.org/spec/expression.html#identity_expressions) compares the identities of two expressions"

we got caught up discussing the mathematical identity property, which is not the property I had in mind in my original post. What I meant was do `a` and `b` in `a is b` point to the same object or not? Saying this, I hope it should be obvious that this is really not about mathematical identities (which is quite messy to define in D, compared to purely functional languages, but that read on below for that) - going back to the original example, `i1 is i2` should be true, despite the fact that the two interfaces `I1` and `I2` don't overlap even in the structural sense. If a function takes a parameter of type `I1`, you really can't call it with an argument of type `I2`. Nevertheless, `i1 is i2` should be true, as they both reference the same object.

So, about my objection about casting `void*`...
> Because you act like I'm cheating by casting to void*, while I consider it a valid expression just like any other.

I don't know about cheating, but I think it defeats your own line of reasoning. Being able to obtain and compare the addresses of objects breaks referential transparency and is big long-standing design problem in D's pure functions - even strongly `pure` functions that accept and return only immutable types can be impure in the mathematical sense, e.g.:

```d
class Num
{
    int x;
    
    @safe pure nothrow @nogc
    this(int x) { this.x = x; }
}

@trusted pure nothrow @nogc
bool areEqual(immutable Num a, immutable Num b)
{
    // In place of a correct implementation (a.x == b.x)
    // let's instead muck around with pointers:
    return cast(const void*) a == cast(const void*) b;
}

@safe pure nothrow
void main()
{
    auto x = new immutable Num(3);
    auto y = new immutable Num(3);
    
    assert(areEqual(x, x));
    assert(areEqual(x, y)); // NOPE
}
```

**Comment by dkorpel** [comment]
> even strongly `pure` functions that accept and return only immutable types can be impure in the mathematical sense, e.g.:

What mathematical definition would consider `areEqual` impure in that example?

**Comment by tgehr** [comment]
The most obvious way in which accessing information about addresses is impure is this:
```d
ulong foo(int x=0)@safe pure=>cast(ulong)&x;

void main(){
	import std.stdio;
	writeln(foo());
	void bar(){
		writeln(foo());
	}
	bar();
}
```

It's a bit more subtle for the `is` operator, but generally speaking, `pure` functions should only depend on the value of their arguments, not on their in-memory representation.

**Pull request closed without merging**